examples/tests/test_01_provider_fastapi.py

```diff
@@ -0,0 +1,142 @@
+"""
+Test the FastAPI provider with Pact.
+
+This module tests the FastAPI provider defined in `src/fastapi.py` against the
+mock consumer. The mock consumer is set up by Pact and will replay the requests
+defined by the consumers. Pact will then validate that the provider responds
+with the expected responses.
+
+The provider will be expected to be in a given state in order to respond to
+certain requests. For example, when fetching a user's information, the provider
+will need to have a user with the given ID in the database. In order to avoid
+side effects, the provider's database calls are mocked out using functionalities
+from `unittest.mock`.
+
+In order to set the provider into the correct state, this test module defines an
+additional endpoint on the provider, in this case `/_pact/provider_states`.
+Calls to this endpoint mock the relevant database calls to set the provider into
+the correct state.
+"""
+
+from __future__ import annotations
+
+from multiprocessing import Process
+from typing import Any, Generator
+from unittest.mock import MagicMock
+
+import pytest
+import uvicorn
+from examples.src.fastapi import app
+from pact import Verifier
+from pydantic import BaseModel
+from yarl import URL
+
+PROVIDER_URL = URL("http://localhost:8080")
+
+
+class ProviderState(BaseModel):
+    """Define the provider state."""
+
+    consumer: str
+    state: str
+
+
+@app.post("/_pact/provider_states")
+async def mock_pact_provider_states(state: ProviderState) -> dict[str, str | None]:
+    """
+    Define the provider state.
+
+    For Pact to be able to correctly tests compliance with the contract, the
+    internal state of the provider needs to be set up correctly. Naively, this
+    would be achieved by setting up the database with the correct data for the
+    test, but this can be slow and error-prone. Instead this is best achieved by
+    mocking the relevant calls to the database so as to avoid any side effects.
```

Comment:
https://docs.pact.io/5-minute-getting-started-guide#testing-the-order-api-provider-project

is it worth being overt about utilising the apps db access layer (db repositories) in the above example?

Reply:
I've added the link at the top. I'm not entirely sure what you're referring to about the apps db access layer...

Reply:
Ahh so 

we have a fake db here in the provider code

https://github.com/pact-foundation/pact-python/blob/f9d2024b19237f1005bee592691ecf237270093e/examples/src/fastapi.py#L35C1-L35C8

and we mock it out on the provider test, with a magic mock

https://github.com/pact-foundation/pact-python/blob/f9d2024b19237f1005bee592691ecf237270093e/examples/tests/test_01_provider_fastapi.py#L113

i suppose in your prod code, you might have an abstraction layer which say says a database with dependency injection, so it could easy use a fake db or a real db to fulfil the db access functions required by the application code (so the main logic of the code, and the tests are agnostic to the fake the database is _real_ or not.

Reply:
this is fine as is anyway, just thinking aloud :)

Reply:
Your production code would have just the production code. This may be abstracted through some layer, or provided by a direct dependency. The idea though is that the test should not change the production code, and instead the test suite does some magic mocking and intercepts the actual calls and returns what is expected.

Do you think it is worth elaborating in the docs?